### PR TITLE
APPSRE-10615 DTP pr-check

### PR DESCRIPTION
**Pull request by fishi0x01**
We cannot catch all spec misconfiguration via json schema. In this PR we add spec validation.

### CODE REVIEW

**Comment by maorfr** [comment]
isn't pr-check == dry-run?

and why wouldn't we do this on non-dry-run as well?

**Comment by fishi0x01** [comment]
@maorfr `validate_token_specs` runs always: `--pr-check`, `--dry-run` and `--no-dry-run`. The idea behind `--pr-check` was to avoid any calls to the OCM API. But I agree that we should be fine with some extra calls and just do `--dry-run`.

File: reconcile/dynatrace_token_provider/validate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                )
+            seen_secrets.add(secret_definition_key)
+
+            seen_tokens: set[str] = set()
+            seen_key_names: set[str] = set()
+            for token in secret.tokens:
+                if len(token.name) > MAX_TOKEN_NAME_LENGTH:
+                    raise TokenNameTooLongError(
+                        f"Token name '{token.name}' is too long. The maximum length is {MAX_TOKEN_NAME_LENGTH} characters."
+                    )
```
Comment (hemslo) [inline reconcile/dynatrace_token_provider/validate.py:47]: size limit can be done by schema `maxLength`

File: reconcile/dynatrace_token_provider/validate.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                        f"A token name must be unique within a secret. Token name '{token.name}' is used more than once in secret '{secret.name}' in token spec '{spec.name}'."
+                    )
+                seen_tokens.add(token.name)
+
+                if token.key_name_in_secret:
+                    if token.key_name_in_secret in seen_key_names:
+                        raise KeyNameNotUniqueInSecretError(
+                            f"A key name must be unique within a secret. Key name '{token.key_name_in_secret}' is used more than once in secret '{secret.name}' in token spec '{spec.name}'."
+                        )
+                    seen_key_names.add(token.key_name_in_secret)
```
Comment (hemslo) [inline reconcile/dynatrace_token_provider/validate.py:60]: `token.name` and `token.key_name_in_secret` used as `secret_key = token.key_name_in_secret or token.name`, should apply combined unique to `secret_key`? https://github.com/app-sre/qontract-reconcile/blob/e41d5c27eec7fc8a181b1d6636d25cfa7c0d891f/reconcile/dynatrace_token_provider/integration.py#L381

File: reconcile/dynatrace_token_provider/validate.py (hunk 1 above)
Comment (fishi0x01) [reply reconcile/dynatrace_token_provider/validate.py]: Oh nice! Created https://github.com/app-sre/qontract-schemas/pull/699 and removed token name length validation.

File: reconcile/dynatrace_token_provider/validate.py (hunk 2 above)
Comment (fishi0x01) [reply reconcile/dynatrace_token_provider/validate.py]: Awesome catch! Added a new test case for this scenario

Comment (hemslo) [reply reconcile/dynatrace_token_provider/validate.py]: Is there a use case for `token.name` to be unique? Or just `secret_key` unique is enough?

Comment (fishi0x01) [reply reconcile/dynatrace_token_provider/validate.py]: Yes `token.name` is used for name in DT API and should be unique.

**Review by hemslo — APPROVED** [review]
LGTM!

**Pull request merged**